File: cpp_euler/0-20/p004.cpp
/* Project Euler - Problem 4
  Find the largest palindrome made from the product of two 3-digit numbers.

  p004.cpp
  g++ p004.cpp -o p4.exe
*/
#include <iostream>

int check_palindrome_6n(int num) { // could also be implemented using bitwise operations (i think)
  return ((num/int(1E5) == num%int(1E1)) &&
          ((num%int(1E5))/int(1E4) == (num/int(1E1))%int(1E1)) &&
          ((num%int(1E4))/int(1E3) == (num/int(1E2))%int(1E1)));
}

int main() {
  int num = 0;
  /*
  std::cout << num/int(1E5) << " == " << num%int(1E1) << std::endl;
  std::cout << (num%int(1E5))/int(1E4) << " == " << (num/int(1E1))%int(1E1) << std::endl;
  std::cout << (num%int(1E4))/int(1E3) << " == " << (num/int(1E2))%int(1E1) << std::endl;
  std::cout << check_palindrome_6n(num) << std::endl;
  */
  for (int i = 999; i > 100; i--){
    for (int j = i; j > 100; j--){
      if (num > i*j) break;
      if (check_palindrome_6n(i*j)){
        std::cout << num << std::endl;
        num = i*j;
      }
    }
    if (num > i*i) break;
  }

  std::cout << num << std::endl; // 906609 (6164 steps)
  return 0;
}

File: rust_euler/0-20/p004.rs
/* Project Euler - Problem 4
  Find the largest palindrome made from the product of two 3-digit numbers.
  
  p004.rs
  rustc p004.rs
  ./p004

  note: this general version is too limited, but it gets the expected result
*/

fn main() {
    let num = 190091;
    println!("num: {}", num);
    println!("is_palindrome(num): {}", is_palindrome(num));

    let result = p004(6);
    println!("result: {}", result);
}

fn p004(n: u32) -> u32{
    let mut num = 0;
    let (min, max) = (u32::pow(10, n/2-1), u32::pow(10, (n-1)/2+1));
    //println!("({}, {})", min, max);
    for i in (min..max).rev() {
        for j in (min..i).rev() {
            if i*j < num {
                break;
                
            }
            if is_palindrome(i*j) {num = i*j;}
        }
        if i*i < num {break;}
    }
    return num;
}

fn is_palindrome(n: u32) -> bool {
    let (mut rev, mut org) = (0,n);
        
    while org > 0 {
        rev = (rev*10) + org%10;
        org/=10;
    }
    rev == n
}
